### PR TITLE
linkage_checker: skip files with incompatible architectures

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Some formulae include these files, and they can't always be removed.
However, they can cause spurious linkage failures, so let's skip them
when checking for linkage. See, for example, faust at
Homebrew/homebrew-core#191308.

### CODE REVIEW

File: Library/Homebrew/os/linux/elf.rb [hunk 1]
```diff
@@ -68,6 +68,12 @@ def arch
     end
   end
 
+  def arch_compatible?(wanted_arch)
+    return true unless elf?
+
+    wanted_arch == arch
```
Comment (Bo98) [inline Library/Homebrew/os/linux/elf.rb:74]: Technically isn't correct for PPC but we can follow up with that.

File: Library/Homebrew/os/mac/mach.rb [hunk 2]
```diff
@@ -155,6 +155,13 @@ def ppc64?
     arch == :ppc64
   end
 
+  def arch_compatible?(wanted_arch)
+    return true unless mach_data.present?
+    return arch == wanted_arch unless universal?
+
+    archs.include?(wanted_arch)
+  end
```
Comment (Bo98) [inline Library/Homebrew/os/mac/mach.rb:163]: I wonder if we should just make this always true on macOS given it's never been problematic? Linux has separate `ld` loaders but macOS does not.

**Review by Bo98 — APPROVED** [review]

**Review by MikeMcQuaid — COMMENTED** [review]
I think rather than silently skipping having this files covered by e.g. `skip_clean` or some other new DSL would be nicer. As-is this seems more likely to increase the times that new formulae bundle e.g. random prebuilt binaries and we don't catch them.

**Comment by Bo98** [comment]
We already do have an audit and an exception list. This is just to fix a regression in the linkage checker after moving from `ldd`.

**Review by Bo98 — APPROVED** [review]

File: Library/Homebrew/os/mac/mach.rb (hunk 2 above)
Comment (carlocab) [reply Library/Homebrew/os/mac/mach.rb]: Done in 6329db9065ae670f0f69454b3e58ec81ba7b3ea3

File: Library/Homebrew/os/linux/elf.rb (hunk 1 above)
Comment (carlocab) [reply Library/Homebrew/os/linux/elf.rb:74]: Going to punt on this!

**Review by MikeMcQuaid — APPROVED** [review]
Thanks for explaining! Happy to merge.

**Pull request merged**